### PR TITLE
[Execution] Move Signature verification to new pipeline phase

### PR DESCRIPTION
**Pull request by sitalkedia**
### Description

- Add a block preparation phase and move the signature verification to the preparation phase. This is to move the signature verification out of the critical execution path - which consumes around 50 ms today for a block size of 25k.
- Remove cloning of transaction block and pass the reference instead to the block executor - this gives us around 2% of performance boost. 

### Test Plan

Tested with the benchmark and this change along with https://github.com/aptos-labs/aptos-core/pull/10388 gives us around 7% performance gain. The TPS 32k to 34k.

### CODE REVIEW

File: aptos-move/aptos-vm/src/aptos_vm.rs
Comment (msmouse) [inline aptos-move/aptos-vm/src/aptos_vm.rs:549]: The loss of type safety in the interfaces (everywhere) is kinda annoying. Can we do better? Maybe introduce something like this? 
```
struct TxnWithValidSig<'a>(&Transaction)
```

```
impl SignatureCheckedTransaction {
    fn expect_valid(&self) -> TxnWithValidSig {
         ...
    }
}
```

File: consensus/src/execution_pipeline.rs
Comment (msmouse) [inline consensus/src/execution_pipeline.rs:34]: don't need to hold this on self?

File: aptos-move/block-executor/src/task.rs
Comment (igor-aptos) [inline aptos-move/block-executor/src/task.rs:33]: I understand that you've renamed it because you needed the type externally, and need to distinguish it. 
But this is going to be a huge conflict with an aggregator branch.

how about having an alias for this module :
// TODO rename to BlockExecutableTransaction after aggregators get merged
type Transaction = BlockExecutableTransaction;

and undoing all the changes in block executor with regards to that?

File: types/src/aggregator.rs
Comment (igor-aptos) [inline types/src/aggregator.rs:15]: hmm, where do you need AggregatorID outside?

File: execution/executor/src/components/chunk_output.rs
Comment (igor-aptos) [inline execution/executor/src/components/chunk_output.rs:248]: cannot there be some trait that will allow these two functions to be merged?

File: execution/executor/src/components/chunk_output.rs
Comment (msmouse) [inline execution/executor/src/components/chunk_output.rs:247]: Man, the duplication is unbearable :trollface: 
Can you do some magic here?

File: types/src/aggregator.rs
Comment (igor-aptos) [reply types/src/aggregator.rs:15]: asking as this ID is getting deprecated and refactored, and shouldn't be needed outside

Comment (sitalkedia) [reply types/src/aggregator.rs:15]: This is needed for the `BlockExecutorTransaction` interface.

File: execution/executor/src/components/chunk_output.rs
Comment (sitalkedia) [reply execution/executor/src/components/chunk_output.rs:248]: Yes, I intended to do this as a follow up, when I move this counter off of critical path. But will add this in this PR.

File: aptos-move/block-executor/src/task.rs
Comment (sitalkedia) [reply aptos-move/block-executor/src/task.rs:33]: Done.

File: aptos-move/aptos-vm/src/aptos_vm.rs
Comment (sitalkedia) [reply aptos-move/aptos-vm/src/aptos_vm.rs]: Too fancy for me.. I moved this function back to the original call to address your paranoia.. PTAL.

File: execution/executor/src/chunk_executor.rs
Comment (sitalkedia) [inline execution/executor/src/chunk_executor.rs:225]: @msmouse, @JoshLind -  Is there any case for keeping signature verification in the state sync path? If not, I will remove this paranoid check..

File: types/src/transaction/signature_verified_transaction.rs
Comment (msmouse) [inline types/src/transaction/signature_verified_transaction.rs:34]: make it this so that paranoid me feels safer :trollface: 
```
    pub fn expect_valid(&self) -> &Transaction {
        match self {
            SignatureVerifiedTransaction::Valid(txn) => txn,
            _ => panic!(),
        }
    }
```

Comment (sitalkedia) [reply types/src/transaction/signature_verified_transaction.rs:34]: Done!

File: execution/executor/src/chunk_executor.rs
Comment (msmouse) [reply execution/executor/src/chunk_executor.rs:225]: I don't think it needs to be checked -- the txns are proven to be part of the chain represented by the ledger info in the proof. The execution mode, without the signature checking, still verifies the VM executes the txns the same with the validator set, and still saves bandwidth transfering the transaction outputs directly -- just one less thing checked.

File: types/src/transaction/signature_verified_transaction.rs
Comment (msmouse) [reply types/src/transaction/signature_verified_transaction.rs:34]: Possible to remove `inner()` to make me feel safeR :trollface:? How do we justify getting the inner txn indifferently not caring if it's signature checked?

Comment (sitalkedia) [reply types/src/transaction/signature_verified_transaction.rs:34]: This is mostly being used for some tests.. let me see if it can be doable..

File: execution/executor/src/chunk_executor.rs
Comment (JoshLind) [reply execution/executor/src/chunk_executor.rs:225]: To clarify, these are the signatures on the individual transactions themselves, right? e.g., verifying the transaction is correctly signed by the sender, etc. If so, this sounds okay to me. But, can we clarify where these checks are being done? i.e., I imagine we're still doing them in mempool, and I imagine that the VM needs to check them, too, right? Otherwise, what if we have a malicious block proposal (e.g., where some transaction signatures are invalid) where would this get caught?

Comment (sitalkedia) [reply execution/executor/src/chunk_executor.rs:225]: > But, can we clarify where these checks are being done? i.e., I imagine we're still doing them in mempool, and I imagine that the VM needs to check them, too, right? Otherwise, what if we have a malicious block proposal (e.g., where some transaction signatures are invalid) where would this get caught?

Yes, currently its being done twice. Once in mempool and once just before the proposal execution (although in a separate pipeline phase). 

> Likewise, what would happen in the case that someone has managed to temporarily compromise enough validators to execute incorrectly signed transactions. Would fullnodes be able to detect this if we remove these checks?

If more than f validators are compromised, then the BFT guarantees doesn't hold anymore.. not sure if FNs detecting invalid signature at that point is worth it..

Comment (JoshLind) [reply execution/executor/src/chunk_executor.rs:225]: Gotcha, thanks! 😄

> If more than f validators are compromised, then the BFT guarantees doesn't hold anymore.. not sure if FNs detecting invalid signature at that point is worth it..

FWIW, I think this point is debatable. One of the big arguments around the need for "verifying read-replicas" (i.e., fullnodes) is that it improves decentralization. Sure, BFT might be compromised in the validator set, but fullnode operators will always be able to detect and call out the divergence. If this no longer becomes the case, how could one detect divergence? I'm not saying don't do it -- it's just worth thinking about the trade-off 😄

Comment (sitalkedia) [reply execution/executor/src/chunk_executor.rs:225]: As discussed offline, let's keep the signature verification for now. We can follow up on this.

File: types/src/transaction/signature_verified_transaction.rs
Comment (sitalkedia) [reply types/src/transaction/signature_verified_transaction.rs]: Removed the inner function.

Comment (sitalkedia) [reply types/src/transaction/signature_verified_transaction.rs]: Please note that we still need the `into_inner` API as we store all txns in the chunk output. This is the same behavior we had previously as well.

**Review by msmouse — APPROVED** [review]

File: types/src/transaction/mod.rs
Comment (zekun000) [inline types/src/transaction/mod.rs:744]: it's confusing to have this and `check_signature`, `SignatureCheckedTransaction` and `SignatureVerifiedTransaction` why do we need to keep the invalid case?

File: types/src/transaction/signature_verified_transaction.rs
Comment (zekun000) [inline types/src/transaction/signature_verified_transaction.rs:74]: nit: it's more idiomatic to impl From

File: types/src/transaction/mod.rs
Comment (sitalkedia) [reply types/src/transaction/mod.rs:744]: @zekun000 - We need to keep the invalid case as we want to notify the mempool/state sync to remove these txns in the commit path.

**Review by zekun000 — APPROVED** [review]

File: types/src/transaction/mod.rs
Comment (zekun000) [reply types/src/transaction/mod.rs:744]: I think it's better to guarantee the block only has valid signature txns (like check in quorum store) but I guess that's for later

Comment (sitalkedia) [reply types/src/transaction/mod.rs:744]: Yes, that is true.. it would be much cleaner to pass only valid sig txn to the executor.. let me add a TODO here and we can follow up on that.

**Pull request merged**
